pkg/audit/verifier.go

```diff
@@ -74,48 +76,75 @@ func (verifier *Verifier) Verify(ctx context.Context, stripe *Stripe, skip map[s
 	shareSize := pointer.GetRemote().GetRedundancy().GetErasureShareSize()
 	bucketID := createBucketID(stripe.SegmentPath)
 
-	// TODO(kaloyan): CreateAuditOrderLimits checks overlay cache if nodes are online and won't return
-	// order limits for offline node. So we miss to record them as offline during the audit
+	var offlineNodes storj.NodeIDList
+	var failedNodes storj.NodeIDList
+	containedNodes := make(map[int]storj.NodeID)
+	sharesToAudit := make(map[int]Share)
+
 	orderLimits, err := verifier.orders.CreateAuditOrderLimits(ctx, verifier.auditor, bucketID, pointer, skip)
 	if err != nil {
 		return nil, err
 	}
 
+	offlineNodes = getOfflineNodes(stripe.Segment, orderLimits)
+
 	shares, nodes, err := verifier.DownloadShares(ctx, orderLimits, stripe.Index, shareSize)
 	if err != nil {
 		return nil, err
 	}
 
-	var offlineNodes storj.NodeIDList
-	var failedNodes storj.NodeIDList
-	containedNodes := make(map[int]storj.NodeID)
-	sharesToAudit := make(map[int]Share)
-
 	for pieceNum, share := range shares {
-		if share.Error != nil {
-			// TODO(kaloyan): we need to check the logic here if we correctly identify offline nodes from those that didn't respond.
-			if share.Error == context.DeadlineExceeded || !transport.Error.Has(share.Error) || ContainError.Has(share.Error) {
-				containedNodes[pieceNum] = nodes[pieceNum]
-			} else {
+		if share.Error == nil {
+			// no error -- share downloaded successfully
+			sharesToAudit[pieceNum] = share
+			continue
+		}
+		if transport.Error.Has(share.Error) {
+			if context.DeadlineExceeded == errs.Unwrap(share.Error) {
+				// dial timeout
 				offlineNodes = append(offlineNodes, nodes[pieceNum])
+				continue
 			}
-		} else {
-			sharesToAudit[pieceNum] = share
+			if codes.Unknown == status.Code(errs.Unwrap(share.Error)) {
+				// dial failed -- offline node
+				offlineNodes = append(offlineNodes, nodes[pieceNum])
+				continue
+			}
+			// unknown transport error
+			containedNodes[pieceNum] = nodes[pieceNum]
+			continue
+		}
+
+		if codes.NotFound == status.Code(errs.Unwrap(share.Error)) {
+			// missing share
+			failedNodes = append(failedNodes, nodes[pieceNum])
+			continue
 		}
+
+		if codes.DeadlineExceeded == status.Code(errs.Unwrap(share.Error)) {
+			// dial successful, but download timed out
+			containedNodes[pieceNum] = nodes[pieceNum]
+			continue
+		}
+
+		// unknown error
```

Comment:
Can we write a log line into the satellite log. That way we can filter all errors that triggered the containment mode and decide if we should exclude them or not.

Reply:
What log severity? Debug, or something else?

Reply:
Log level debug should be fine. Please also include the storage nodeID just in case we get some strange results and want to request the storage node logs.

Reply:
Done.